src/common/pico_time/time.c

```diff
@@ -439,26 +444,35 @@ bool best_effort_wfe_or_timeout(absolute_time_t timeout_timestamp) {
         return time_reached(timeout_timestamp);
     } else {
         alarm_id_t id;
-        id = add_alarm_at(timeout_timestamp, sleep_until_callback, NULL, false);
-        if (id <= 0) {
-            tight_loop_contents();
+        // note that as of SDK2.0.0 calling add_alarm_at always causes a SEV. Whwat we really
+        // want to do is cause an IRQ at the specified time in the future if there is not
+        // an IRQ already happening before then. The problem is that the IRQ may be happening on the
+        // other core, so taking an IRQ is the only way to get the state protection.
+        //
+        // Therefore, we make a compromise; we will set the alarm, if we won't wake up before the right time
+        // already. This means that repeated calls to this function with the same timeout will work correctly
+        // after the first one! This is fine, because we ask callers to use a polling loop on another
+        // event variable when using this function
+        if (ta_wakes_up_on_or_before(alarm_pool_get_default()->timer, alarm_pool_get_default()->timer_alarm_num,
```

Comment:
What happens if the other core (or a preemptive IRQ on the current core) modifies this default timer alarm in the middle of this function call?  This call here doesn't look atomic, and the timer may be updated (cancelled, moved further into the future) before the `__wfe()` just below, which would make the `__wfe()` wait longer than the requested timeout?

Reply:
ugh true; that's what i get for trying to write code with COVID!

Reply:
although actually the circumstances where it could not get woken up are really just the other core canceling an alarm (or adding one further out in the future)... i think adding a SEV at the end of the alarm (IRQ) handler should solve that.

Reply:
> that's what i get for trying to write code with COVID!

Oh no!  You should just rest.

> although actually the circumstances where it could not get woken up are really just the other core canceling an alarm (or adding the first one further out in the future)... i think adding a SEV at the end of the alarm (IRQ) handler should solve that.

I think the alarm needs to guarantee that, if its IRQ gets enabled then:
1. when the IRQ fires it will do a `__sev()`
2. the alarm time will never be delayed or cancelled, it can only ever be brought forward (earlier) in time

Reply:
Yeah actually I think I misunderstood what you meant when I read it yesterday… when the irq handler sets a new time, it should never make it later than the current time - it should always slow the previous target to happen first

Reply:
> when the irq handler sets a new time, it should never make it later than the current time

Yes.

Also, it should never cancel/disable an IRQ if it's already set up.

Reply:
Yeah, sorry, this is in the context of the alarm for the alarm pool which is never disabled any more - it is doomed to fire every 2^32 us